### PR TITLE
Tfsec

### PR DESCRIPTION
**Pull request by BenRamchandani**
Replaces #2.

Fix/ignore the tfsec warnings in the current code. Ignored warnings we should consider looking at, opinions welcome:

* ~Terraform state bucket~
  * ~Enabling (server side) bucket encryption~
    * ~With a customer managed key~
  * ~Enabling access logging (which would go to another S3 bucket)~
* ldaps_crl_and_certs S3 bucket
  * Same as above
  * Does this have secrets in or is it just public keys and a CRL?
    * Edit: Treating as non-secret
* ~Secrets manager~
  * ~Use a CMK - not convinced this is useful at the moment, but we will probably want to do this eventually with different keys for different services~
* AD Management Server
  * Limiting egress - probably something to think about when we set up an egress proxy - what's the ticket number for that?

~It got me to add descriptions for the `domain_controllers_to_ca_` security group rules @hugh-emerson do you know what the ports are for?~

Note that this PR includes changes to the backend state bucket setup, which will presumably need to be applied manually.

I've added a tfsec CI job to PRs only, do we want it for the main branch too, and if so should it block deploys or be allowed to fail?

I can't run the terraform without an AWS key, so I'll let the CI check it.

### CODE REVIEW

File: terraform/modules/active_directory/security_groups.tf [hunk 1]
```diff
@@ -18,11 +18,14 @@ resource "aws_security_group" "ad_management_server" {
...
     protocol  = "-1"
 
+    # tfsec:ignore:aws-vpc-no-public-egress-sgr
     cidr_blocks = ["0.0.0.0/0"]
+    description = "Allow all egress"
   }
 }
 
 resource "aws_security_group_rule" "domain_controllers_to_ca_1" {
+  description       = "Directory Services to CA port 135" # TODO:tfsec what is this for, NTP?
```
Comment (hugh-emerson) [inline terraform/modules/active_directory/security_groups.tf:28]: Either for remote group policy updates or requesting a certificate
Not super sure but it was in AWS's guide and LDAPS didn't work without it

File: terraform/backend/main.tf [hunk 2]
```diff
@@ -2,20 +2,34 @@ provider "aws" {
   region = "eu-west-1"
 }
 
+# TODO:tfsec Enable bucket encryption
+# TODO:tfsec Consider enabling access logging
```
Comment (hugh-emerson) [inline terraform/backend/main.tf:6]: Is there any reason not to enable bucket encryption now?

Yes you're right that I applied this manually. But I guess we could move it to a module and let test/staging/prod include it themselves now.  Just need to import the existing resources

**Review by hugh-emerson — CHANGES_REQUESTED** [review]
Looks good but can you enable bucket encryption and using a CMK to encrypt the secrets as part of this ticket or is there any reason we wouldn't want to do either of those?

Sounds like using a CMK is a pretty unimportant, but also pretty straightforward?
https://aquasecurity.github.io/tfsec/v0.63.1/checks/aws/ssm/secret-use-customer-key/

**Comment by hugh-emerson** [comment]
I've added "tfsec PR commenter" as a required check for PRs: https://github.com/communitiesuk/delta-common-infrastructure/settings/branch_protection_rules/28146899

**Comment by hugh-emerson** [comment]
Looking at how it gets used, the ldaps_crl_and_certs bucket might only be for CRL, in which case it's not sensitive, but tfstate should be encrypted.

File: .github/workflows/pull_requests.yml [hunk 3]
```diff
@@ -31,3 +31,14 @@ jobs:
...
       access_key: "${{secrets.access_key_prod}}"
       secret_key: "${{secrets.secret_key_prod}}"
+  tfsec:
+    name: tfsec PR commenter
+    runs-on: ubuntu-latest
+    steps:
+      - name: Clone repo
+        uses: actions/checkout@master
+      - name: tfsec
+        uses: tfsec/tfsec-pr-commenter-action@main
```
Comment (hugh-emerson) [inline .github/workflows/pull_requests.yml:41]: Looks like `tfsec/tfsec-pr-commenter-action` just redirects to `aquasecurity/tfsec-pr-commenter-action` so I think it'd be clearer just to use the latter?

File: terraform/modules/active_directory/security_groups.tf (hunk 1 above)
Comment (BenRamchandani) [reply terraform/modules/active_directory/security_groups.tf:28]: Cool, I've left it vague for now

File: terraform/backend/main.tf (hunk 2 above)
Comment (BenRamchandani) [reply terraform/backend/main.tf]: Encryption and logging enabled now :+1:

I think just keep it manual, it feels wrong to import the state bucket into itself.

File: .github/workflows/pull_requests.yml (hunk 3 above)
Comment (BenRamchandani) [reply .github/workflows/pull_requests.yml]: Sounds sensible, done

**Comment by BenRamchandani** [comment]
I've also added a separate workflow for running tfsec on the main branch. For now it always uses the latest version of tfsec, so will probably break sometimes, and we don't want that blocking releases.

**Comment by hugh-emerson** [comment]
Looks good. The ticket number for sending egress via Squid is DT-23, but no real need to add it to the comment. Pretty intuitive to revisit the security group at that point.

**Pull request merged**